### PR TITLE
feat(kuma-cp) user specified load balancer type per route

### LINKED ISSUE #1030: Support for consistent hash based load balancing
### Summary

The feature is [supported by Istio](https://istio.io/latest/docs/reference/config/networking/destination-rule/#LoadBalancerSettings) and it can be useful when deploying stateful services like web servers with in-memory session storage.

### PR DESCRIPTION
**Pull request by nickolaev**
### Summary

Extending the TrafficRoute to be able to specify the LoadBalancer type, reflecting Envoy's capabilities.
Supported types:
 * Round-robin
 * Least request - can set choice count
 * Ring hash - can set hash function, min/max ring size
 * Random
 * Maglev

A sample Universal resource would look like this:
```
type: TrafficRoute
name: route-all-default
mesh: default
sources:
  - match:
      kuma.io/service: '*'
destinations:
  - match:
      kuma.io/service: '*'
conf:
  loadBalancer:
    roundRobin: {}
  split:
    - destination:
        kuma.io/service: '*'
      weight: 100
```

Where the `loadBalancer` field can have the following values
```
  loadBalancer:
    leastRequest:
      choiceCount: 32
```
```
  loadBalancer:
    ringHash:
      hashFunction: "MURMUR_HASH_2"
      minRingSize: 64
      maxRingSize: 1024
```
```
  loadBalancer:
    random: {}
```
```
  loadBalancer:
    maglev: {}
```

### Issues resolved

Fix #1030 

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/351)

### CODE REVIEW

File: pkg/core/resources/apis/mesh/traffic_route_validator.go [hunk 1]
```diff
@@ -43,5 +44,28 @@ func (d *TrafficRouteResource) validateConf() (err validators.ValidationError) {
...
+
+	switch lb.LbType.(type) {
+	case *mesh_proto.TrafficRoute_LoadBalancer_LeastRequest_:
+
+	case *mesh_proto.TrafficRoute_LoadBalancer_RingHash_:
+		lbConfig := lb.GetRingHash()
+		switch lbConfig.HashFunction {
+		case "XX_HASH", "MURMUR_HASH_2":
+		default:
+			root := validators.RootedAt("conf.loadBalancer.ringHash")
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/apis/mesh/traffic_route_validator.go:66]: `conf.loadBalancer.ringHash.hashFunction`

File: pkg/xds/envoy/clusters/v2/lb_configurer.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+type LbConfigurer struct {
+	Lb *mesh_proto.TrafficRoute_LoadBalancer
+}
+
+var _ ClusterConfigurer = &LbConfigurer{}
+
+func (e *LbConfigurer) Configure(c *envoy_api.Cluster) error {
+	// default ot Round Robin
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/clusters/v2/lb_configurer.go:20]: `ot` -> `to`

File: pkg/xds/envoy/clusters/v2/lb_configurer_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            connectTimeout: 5s
+            edsClusterConfig:
+              edsConfig:
+                ads: {}
+            name: backend
+            type: EDS`,
+		}),
+		Entry("least request", testCase{
+			clusterName: "backend",
+			lb: &v1alpha1.TrafficRoute_LoadBalancer{
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/clusters/v2/lb_configurer_test.go:52]: nit: alias `v1alpha1` to `mesh_proto`

**Review by jakubdyszkiewicz — APPROVED** [review]
some nits, looks great!

**Comment by jakubdyszkiewicz** [comment]
(and it's `feat(kuma-cp)`)

**Comment by lobkovilya** [comment]
I think it might be confusing that `loadBalancer ` field is located on the top level of `conf`. It feels like this load balancer will be used to balance traffic across `destinations`, but in fact, we don't balance across the destinations, moreover, we have a NO_FALLBACK policy between these destinations. 

I'd suggest renaming this top level `loadBalancer ` to `defaultLoadBalancer` and have `loadBalancer` field for every destination:
```yaml
type: TrafficRoute
name: route-all-default
mesh: default
sources:
  - match:
      kuma.io/service: '*'
destinations:
  - match:
      kuma.io/service: '*'
conf:
  defaultLoadBalancer:
    roundRobin: {}
  split:
    - destination:
        kuma.io/service: backend
        version: 1
      weight: 90
      loadBalancer: 
        maglev: {}
    - destination:
        kuma.io/service: backend
        version: 2
      weight: 5
      loadBalancer:
        random: {}
    - destination:
        kuma.io/service: backend
        version: 3
      weight: 5
```

What do you think about it?

**Comment by jakubdyszkiewicz** [comment]
@lobkovilya 
LB is placed on cluster, how would you like to place it on individual splits?

**Comment by lobkovilya** [comment]
Sorry, I was confused by this [line](https://github.com/kumahq/kuma/pull/1402/files#diff-82033f55666d41e5ff6d6fc36c05b66ef85713bf616d36c080df74e02665d01aR276): 
```go
subset.Lb = route.Spec.GetConf().GetLoadBalancer()
```
nevermind

**Comment by nickolaev** [comment]
@mergifyio update

**Pull request merged**